plugins/algoltek-usb/fu-algoltek-usb-device.c

```diff
@@ -0,0 +1,633 @@
+/*
+ * Copyright (C) 2023 Ling.Chen <ling.chen@algoltek.com.tw>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include "fu-algoltek-usb-common.h"
+#include "fu-algoltek-usb-device.h"
+#include "fu-algoltek-usb-firmware.h"
+// #include "fu-algoltek-usb-struct.h"
+
+struct _FuAlgoltekUsbDevice {
+	FuUsbDevice parent_instance;
+};
+
+G_DEFINE_TYPE(FuAlgoltekUsbDevice, fu_algoltek_usb_device, FU_TYPE_USB_DEVICE)
+
+static void
+fu_algoltek_usb_device_to_string(FuDevice *device, guint idt, GString *str)
+{
+	FU_DEVICE_CLASS(fu_algoltek_usb_device_parent_class)->to_string(device, idt, str);
+}
+
+static guint8
+checkSum(GByteArray *byteArray)
+{
+	guint8 result = 0;
+	for (guint32 i = 0; i < byteArray->len; i++) {
+		result += (guint8)(byteArray->data[i] & 0xff);
+	}
+	result = (guint8)(~result + 1);
+	return result;
+}
+
+static GByteArray *
+fu_algoltek_device_CMD02(FuAlgoltekUsbDevice *self, int address, GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev(FU_USB_DEVICE(self));
+	g_autoptr(GByteArray) buf = g_byte_array_new();
+
+	/* first byte is length */
+	fu_byte_array_set_size(buf, 11, 0x0);
+	buf->data[0] = 5;
+	buf->data[1] = ALGOLTEK_USB_REQUEST_CMD02;
+	buf->data[2] = ((address >> 8) & 0xFF);
+	buf->data[3] = (address & 0xFF);
+	buf->data[10] = checkSum(buf);
```

Comment:
If you added something like this in the .rs file:

	#[repr(u8)]
	enum AlgoltekCmd {
	    // others here
	    SetAddress = 0x6,
	    // more
	}
	#[derive(New)]
	struct AlgoltekCmdAddress {
	    hdrlen: u8 == 5,
	    cmd: AlgoltekCmd == SetAddress,
	    address: u16le,
	    reserved: [u8; 6],
	    checksum: u8,
	}

you can do something like this:

	g_autoptr(GPtrArray) st = fu_struct_algoltek_cmd_address_new()
	fu_struct_algoltek_cmd_address_set_address(st, address);
	fu_struct_algoltek_cmd_address_set_checksum(st, ~fu_sum8(st->data, st->len) + 1);

... and I think that's much more readable.

Reply:
Thanks for your advice.
I compared the transfer packet in each command, but no single format packet matched all commands.
So, I still keep the original way of distributing packets.
However, I split the control transfer function in each command method into a new function. Hope this will make the code clearer and shorter.

Reply:
Thanks for your suggestions.
About the commands ,I moved commands to a new enumeration in the .rs file.